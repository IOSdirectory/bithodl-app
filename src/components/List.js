import React, { Component } from 'react';
import { StyleSheet, View, ScrollView } from 'react-native';
import PropTypes from 'prop-types';
import Row from '../components/Row';

const BORDER_RADIUS = 6;

const styles = StyleSheet.create({
  list: {
    alignSelf: 'stretch',
    backgroundColor: '#FEFCFB',
    minHeight: 426,
    borderRadius: BORDER_RADIUS,
    shadowColor: '#000000',
    shadowRadius: 4,
    shadowOpacity: 0.25,
    shadowOffset: {
      height: 0,
      width: 0
    },
    marginTop: 213,
    marginBottom: 10
  }
});

export default class List extends Component {
  _getItems() {
    const list = Object.values(this.props.items || {});

    list.sort((a, b) => {
      if (a.unlockAt < b.unlockAt) {
        return -1;
      }

      if (a.unlockAt > b.unlockAt) {
        return 1;
      }

      return 0;
    });

    return list;
  }

  _renderRows(items) {
    const firstStyle = {
      borderTopLeftRadius: BORDER_RADIUS,
      borderTopRightRadius: BORDER_RADIUS
    };

    const lastStyle = {
      borderBottomLeftRadius: BORDER_RADIUS,
      borderBottomRightRadius: BORDER_RADIUS
    };

    const rows = items.map((item, index) => {
      let style;

      if (index === 0) {
        style = firstStyle;
      } else if (index === items.length - 1) {
        style = lastStyle;
      }

      return (
        <Row item={item} key={item.id} style={style} onShowItem={this.props.onShowItem} />
      );
    });

    return rows;
  }

  render() {
    const items = this._getItems();

    return (
      <ScrollView showsVerticalScrollIndicator={false} style={this.props.style}>
        <View style={styles.list}>
          {this._renderRows(items)}
        </View>
      </ScrollView>
    );
  }
}

List.propTypes = {
  style: PropTypes.any,
  items: PropTypes.object,
  onShowItem: PropTypes.func
};
