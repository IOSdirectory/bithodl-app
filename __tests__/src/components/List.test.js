import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import List from '../../../src/components/List';

describe('List', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <List />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('#_renderRows()', () => {
    var onShowItem;
    var list;
    var items;
    var returnValue;

    beforeEach(() => {
      onShowItem = jest.fn();

      list = new List();
      list.props = { onShowItem };

      items = [
        { id: '40fd1773-2582-4587-b105-de91534dac5f' },
        { id: '25c69897-0f0c-460f-b640-fb74e8f5f1cd' },
        { id: 'ab5c2461-f04a-4f33-a643-4dd271deacb5' }
      ];

      returnValue = list._renderRows(items);
    });

    it('accepts one argument', () => {
      const length = list._renderRows.length;
      expect(length).toBe(1);
    });

    it('returns an array', () => {
      const returnValue = list._renderRows(items);
      expect(Array.isArray(returnValue)).toBe(true);
    });

    describe('the returned array', () => {
      it('has the length of "items"', () => {
        expect(returnValue.length).toBe(items.length);
      });

      describe('each item in the array', () => {
        it('is of type "Row"', () => {
          returnValue.forEach((row) => {
            expect(typeof row).toBe('object');
            expect(typeof row.type).toBe('function');
            expect(row.type.name).toBe('Row');
          });
        });

        it('has key assigned to the id of the item', () => {
          returnValue.forEach((row, index) => {
            expect(row.key).toBe(items[index].id);
          });
        });

        it('has property "item" assigned to the item', () => {
          returnValue.forEach((row, index) => {
            expect(row.props.item).toBe(items[index]);
          });
        });

        it('has property "onShowItem" assigned to props.onShowItem', () => {
          returnValue.forEach((row) => {
            expect(row.props.onShowItem).toBe(onShowItem);
          });
        });
      });
    });
  });
});
