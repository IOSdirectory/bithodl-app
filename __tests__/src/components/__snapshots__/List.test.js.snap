// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders correctly 1`] = `
<RCTScrollView
  showsVerticalScrollIndicator={false}
  style={undefined}
>
  <View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "#FEFCFB",
          "borderRadius": 6,
          "marginBottom": 10,
          "marginTop": 213,
          "minHeight": 426,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 4,
        }
      }
    />
  </View>
</RCTScrollView>
`;
